Polish focus styles

jasmussen opened the pull request:
This is a rather mammoth PR to improve focus styles across the board. Outside of the switch styles, which need more work and should be addressed separtely, this fixes #3382 and fixes #3039.

The impetus for this design is that we use a lot of "naked" components. Movers, ellipsis menus, icon-only buttons, which when given the traditional focus style appear immaterial. By giving a considered style and background, it gives solidity to the button.

Because it's a mammoth PR that touches everything at once, I've been able to reduce and trim a ton of styles, move many of the focus styles into mixins for reusability, and I've overall been able to do a ton of consistency work using SCSS variables. Border radii, menu item paddings, focus styles and alignments should all improve as of this. 

On the downside, we should test this PR pretty thoroughly.

**Edit**: Forgot to mention — this PR also both unifies and changes the _hover_ style. It's now _very dark gray_ everywhere, except for formatting controls where there's still that 1px border. 

I like this, I think it works, but there's not a lot of visual contrast between neutral and hover states now. I don't think it's a problem, but I wanted to surface it in case others do.

## Screenshots:

![screen shot 2017-11-09 at 11 38 22](https://user-images.githubusercontent.com/1204802/32601566-5efc6340-c543-11e7-9e25-3495374b829c.png)
![screen shot 2017-11-09 at 11 38 28](https://user-images.githubusercontent.com/1204802/32601568-6068e460-c543-11e7-82c8-65b3782b3abc.png)
![screen shot 2017-11-09 at 11 38 32](https://user-images.githubusercontent.com/1204802/32601571-61f317f6-c543-11e7-86f9-63042a7295b8.png)
![screen shot 2017-11-09 at 11 38 38](https://user-images.githubusercontent.com/1204802/32601575-638c3b6a-c543-11e7-83d8-2578bb6dc309.png)
![screen shot 2017-11-09 at 11 38 42](https://user-images.githubusercontent.com/1204802/32601579-663596d6-c543-11e7-839d-edc61f4d1ae4.png)
![screen shot 2017-11-09 at 11 38 46](https://user-images.githubusercontent.com/1204802/32601583-6835a6a6-c543-11e7-97c6-20d60cbb7aa2.png)
![screen shot 2017-11-09 at 11 38 54](https://user-images.githubusercontent.com/1204802/32601585-69b427e6-c543-11e7-8898-e6a227a6ea24.png)
![screen shot 2017-11-09 at 11 38 58](https://user-images.githubusercontent.com/1204802/32601588-6bd39d68-c543-11e7-8b42-e8c8b2a9fd82.png)
![screen shot 2017-11-09 at 11 39 02](https://user-images.githubusercontent.com/1204802/32601590-6ceafec6-c543-11e7-82f2-72fc793bf2dd.png)
![screen shot 2017-11-09 at 11 39 07](https://user-images.githubusercontent.com/1204802/32601591-6e4c3ca8-c543-11e7-803d-bd6af612b426.png)
![screen shot 2017-11-09 at 11 39 14](https://user-images.githubusercontent.com/1204802/32601596-709a58be-c543-11e7-97ef-bde9b7efe4eb.png)
![screen shot 2017-11-09 at 11 39 21](https://user-images.githubusercontent.com/1204802/32601597-715ed838-c543-11e7-97c9-c8fc231cd975.png)

Commit:
Focus style work.

This is a work in progress. Pushing so I can continue work from the coworking space.

Commit:
Polish work on the focus styles

- Improve toggles, search fields
- Cleaner separation of code for borderradii

Commit:
Polish formatting buttons.

Also improve radius consistency further.

Commit:
Fix issue with block-fixed toolbar jogging the block.

Commit:
Fix dropdown menus and popovers.

Fixes #3382.

Commit:
Fix Chrome specific border radius regression in tabs.

youknowriad commented:
Awesome work. It's working pretty well globally.

Two questions:

 - The in-between blocks inserter has a different style, maybe you forgot about it?
 - Do we really need the "double" border on focus styles: the inner white and the black border. I think it may look better if we drop the white. Personal opinion though. (Maybe aside when we have both toggled and focus style at the same time)

youknowriad commented on editor/block-switcher/index.js:
```diff
@@ -110,9 +110,7 @@ function BlockSwitcher( { blocks, onTransform } ) {
 								} }
 								className="editor-block-switcher__menu-item"
 								icon={ (
-									<span className="editor-block-switcher__block-icon">
```
The block icon is not always a dashicon, this wrapper was necessary to correctly add the spacing between the icon and text when we have custom elements as icons. Not saying we shouldn't remove it but it was there on purpose and we should make sure it doesn't break. #3238 

I think you're targetting `svg` but block icons are not always `svg`, they can be anything.

Commit:
Style in-between inserter and other polish

- Styles the in-between inserter focus style
- Fix some issues with the block-docked toolbar stacking left-borders
- Restore the icon span to the transformer, and re-do style neutralization.

jasmussen commented:
> The in-between blocks inserter has a different style, maybe you forgot about it?

I did, thanks!

> Do we really need the "double" border on focus styles: the inner white and the black border. I think it may look better if we drop the white. Personal opinion though. (Maybe aside when we have both toggled and focus style at the same time)

I think we do, as there are apparently some very specific requirements around contrast between _shapes_ and backgrounds, and the background itself doesn't count as a shape. 

But since we are moving this all to mixins, adjusting these in the future if we find that we can, should be easier to do in one place instead of 20 different scss files.

I restored the span and fixed things with a different approach. Good catch, Riad, thank you!

jasmussen commented:
Rebased. I think this is ready for a final pair of eyes. CC: @mtias @karmatosed

Commit:
Remove word count specific style.

jasmussen commented:
Got a thumbs up in chat. Merging.
